### ISSUE TITLE
Couldn't locate libffmpegsumo.so during installation. Audio notifications will be disabled

### ISSUE THREAD

**Issue by baniobloom** [plaidchat/plaidchat#100]
I'm using Debian Jessie amd64.  I installed plaidchat with npm and got the error:
Couldn't locate libffmpegsumo.so during installation. Audio notifications will be disabled

However I already have chromium installed and the libffmpegsumo file is here:
/usr/lib/chromium/libffmpegsumo.so

**Comment by twolfson** [comment]
The current solution only looks for `/opt/google/chrome`. Can you try updating `postinstall.sh` to attempt to resolve/copy from `/usr/lib/chromium`, verify that audio notifications work, and then submit a PR?

https://github.com/plaidchat/plaidchat/blob/v2.5.1/postinstall.sh#L7-L8

**Comment by subjectdenied** [comment]
please also test for /opt/google/chrome-unstable (chrome-dev builds)

**Comment by rosshadden** [comment]
:+1: for `chrome-unstable`.  People also use `chrome-beta`, so we need to account for all of those.

**Comment by Mte90** [comment]
done a pr :-)

**Comment by twolfson** [comment]
The original requested Chromium fix has been landed in `2.10.0`. I will be submitting a follow up PR to take care of the straggling `chrome-beta` and `chrome-unstable`

**Comment by twolfson** [comment]
As promised, we have added support for `chrome-beta` and `chrome-unstable` and it has been released in `2.11.0`. Thanks everyone for their contributions to this issue =)

**Comment by viaductit** [comment]
it looks like neither chrome-stable nor chrome-beta/chrome-unstable are shipping with libffmpegsumo.so any more (i double-checked, this isn't an arch-linux thing, google isn't shipping it within their official packages). i tried to use the .so from some other nw.js apps, however while plaidchat's setup-script does accept them, audio doesn't work with it. do you have any idea where to get a working .so for plaidchat, and would it be possible to implement audio without using it?

**Comment by subjectdenied** [comment]
it looks like neither chrome-stable nor chrome-beta/chrome-unstable are shipping with libffmpegsumo.so any more (i double-checked, this isn't an arch-linux thing, google isn't shipping it within their official packages). i tried to use the .so from some other nw.js apps, however while plaidchat's setup-script does accept them, audio doesn't work with it. do you have any idea where to get a working .so for plaidchat, and would it be possible to implement audio without using it?

sry for double-posting, was logged in with wrong account ;-)

**Comment by twolfson** [comment]
If Chrome is no longer shipping `libffmpegsumo.so`, then they won't be able to play mp3's. Can you verify that you can play mp3's from this test page in Chrome?

http://hpr.dogphilosophy.net/test/

**Comment by subjectdenied** [comment]
well i am able to play the mp3-file from the page you mentioned, however i think this is done differently in chrome now. cause i do not have a libffmpegsumo.so inside /opt/google/chrome-unstable/ for an example, and i'm pretty sure this is the only place chrome libs are placed (at least in arch)

**Comment by twolfson** [comment]
Hmmm, damn. For my sanity, can you run the following?

```bash
find / -name '*ffmpegsumo.so*' 2> /dev/null
```

**Comment by subjectdenied** [comment]
/opt/Gitter/linux64/libffmpegsumo.so
/opt/MessengerForDesktop/libffmpegsumo.so
/opt/nw.js/libffmpegsumo.so
/opt/visual-studio-code/libffmpegsumo.so
/usr/share/atom/libffmpegsumo.so
/usr/share/spotify/libffmpegsumo.so
/usr/lib/node-webkit/libffmpegsumo.so
/usr/lib/node_modules/plaidchat/node_modules/nw/nwjs/libffmpegsumo.so
/usr/lib/node_modules/electron-prebuilt/dist/libffmpegsumo.so

if you wonder, the current non working libffmpegsumo.so plaidchat is using, was copied into a /opt/google/chrome/ directory before i installed plaidchat, so it is caught by the install-script, i removed the dummy directory afterwards

**Comment by twolfson** [comment]
Ah, damn. Unfortunately, the beta of the latest `nw.js` isn't out on `npm-installer` yet either so we prob can't trust it.

Thankfully, the `libffmpegsumo.so` in `electron-prebuilt` peaked my interest. After some reseearch, I found a node module that offers prebuilt binaries of `libffmpegsumo.so`. We should explore using that module over copying from Google Chrome/Chromium:

https://github.com/eugeneware/ffmpeg-static

**Comment by twolfson** [comment]
Damn, never mind. The module seems to be an executable itself, not the shared object library we want =(

**Comment by twolfson** [comment]
After downloading the `0.13` alpha of `nw.js`, I discovered a `libffmpegsumo.so` in its bundle so it looks like this step is on its way out. Unfortunately, the beta seems unstable and I am having trouble running `plaidchat`.

https://github.com/nwjs/nw.js/tree/69daf184db463c060709299129d79d36c4afbb65#downloads

For now, I think our recommendation should be to wait to see when the next `0.13` stable release is coming out. It seems to be every 3 months based on the release tags, which means we are coming up shortly as the last minor release was March 4th.

https://github.com/nwjs/nw.js/releases

**Comment by twolfson** [comment]
It looks like waiting has paid off. `nw.js@0.12.3` was released yesterday and seems like it's bundled with `libffmpegsumo.so`

https://groups.google.com/forum/#!msg/nwjs-general/hhXCS4aXGV0/TUQmcu5XDwAJ

http://dl.nwjs.io/v0.12.3/chromedriver-nw-v0.12.3-linux-x64.tar.gz

However, there is no release on `npm-installer` yet so I will open a new issue.

**Comment by subjectdenied** [comment]
https://github.com/aprilorange/nwjs

would this help?

**Comment by twolfson** [comment]
That works as well but the maintainers of `npm-installer` have a reasonable response time so we are probably fine sitting tight for now.

http://issuestats.com/github/nwjs/npm-installer

**Comment by twolfson** [comment]
`npm-installer` released its latest version but it looks like the bundled `libffmpegsumo.so` is failing to play audio notifications in `plaidchat`. When I copy over my `libffmpegsumo.so` from `electron-prebuilt`, it starts working.

I am going to file an issue so this is properly resolved for everyone rather than us working around it.

**Comment by twolfson** [comment]
For reference, my work on using the new `nw.js` can be found here:

https://github.com/twolfson/plaidchat/tree/c5ad7ae0fb35a0b8ee9199777481849addfda5e4

**Comment by rdtsc** [comment]
@twolfson: Further to our discussion in: nwjs/nw.js#3723 -- Unless you're willing to mess with something like HTTPFS, I don't think there's much you can do without downloading the entire set of pre-built binaries (in a single ZIP), either directly or via `electron-prebuilt` and then cherry-picking your shared objects.

Alternatively, you can create a secondary project on GitHub, for example: `plaidchat-dep`, and use the [Releases](https://github.com/blog/1547-release-your-software) feature to store shared objects for each platform/architecture you support. The master branch should contain:
- A basic readme stating that this is a support repo and is not meant for direct user consumption.
- A basic file registry, in JSON, for example. 

```text
plaidchat-dep (master)
├── index.json
└── readme.md
```
The `index.json` file registry is responsible for mapping your local dependencies to their remote counterparts, for example:
```json
{
  "v1.0.0":
  {
    "linux-x86": 
    [
      {
        "filename": "libffmpegsumo.so", 
        "binary": "https://github.com/plaidchat/plaidchat-dep/releases/download/v1.0.0/libffmpegsumo-x86.so", 
        "sha256": "421c76d77563afa1914846b010bd164f395bd34c2102e5e99e0cb9cf173c1d87", 
        "installPath": "./"
      }
    ], 
    
    "linux-x64":
    [
      {
        "filename": "libffmpegsumo.so", 
        "binary": "https://github.com/plaidchat/plaidchat-dep/releases/download/v1.0.0/libffmpegsumo-x64.so", 
        "sha256": "6c855792af732daf809592e38920c58a886e3df741bac7bbde98a2e6c1c0925c", 
        "installPath": "./"
      }
    ]
  }
}
```
The actual binary assets are hosted in the project's Releases section, for example:

![concept](https://cloud.githubusercontent.com/assets/6773893/9028392/1f97c0aa-3945-11e5-842e-b573fa99b02e.png)

Next, on `postinstall`, run a script that consumes `https://raw.githubusercontent.com/plaidchat/plaidchat-dep/master/index.json` and downloads the required NW.js dependencies for the current platform if they don't already exist or if their hash differs from what's in the remote registry.

Seems to me like the entire process can be automated without much trouble. Before publishing to NPM, your script should:
- Fetch all relevant copies of Electron's binaries in ZIP format.
- Decompress and save the relevant shared object files.
- Update the JSON registry.
- Upload the shared objects as a Release assets via [GitHub's API](https://developer.github.com/v3/repos/releases/#upload-a-release-asset).
- Commit and push the file registry to GitHub.

Also, to quote NW's [wiki](https://github.com/nwjs/nw.js/wiki/Using-MP3-&-MP4-%28H.264%29-using-the--video--&--audio--tags.) again:
> **Consult a lawyer if you do not understand the licensing constraints and require patented media formats in your application.**

**Comment by twolfson** [comment]
I am not sure why this discussion is being brought back to `plaidchat`. Our problem isn't specific to `plaidchat`; it's generic to all mp3 support on nw.js projects.

Building a `plaidchat-dep` repository is short sighted because it is specific to `plaidchat`. It is inevitable that there will be another project which runs into the same problem of not being able to use a prebuilt `libffmpegsumo.so` from Google Chrome (which is still documented in the nw.js wiki). Hence, it's more of a platform problem and this should be discussed in `nw.js`.

We are already aware of the GPL copyleft required by Google Chrome's `libffmpegsumo.so` and document it in our README.

https://github.com/plaidchat/plaidchat/tree/v2.11.0#license

**Comment by rdtsc** [comment]
@twolfson: You are right. I was under the impression that you were looking for a project-specific solution. I'll go back to: nwjs/nw.js#3723

**Comment by twolfson** [comment]
I am reopening this issue since it now exists in the form of Google Chrome not having the binary. I am going to close it shortly again with an `electron's libffmpegsumo.so` copy. But for now, it should be open.

**Comment by twolfson** [comment]
The patch PR was landed in #129. Closing this issue
